srk/src/solvablePolynomial.ml

```diff
@@ -1955,25 +1947,125 @@ end
 
 module I = Polynomial.Rewrite
 
+module Id = Polynomial.Ideal
+
 module SolvablePolynomialLIRR = struct
+
+  type pre_t = 
+    {
+      ideal : TransitionIdeal.t
+    ; witness : TransitionIdeal.solvable_polynomial
+    }
+
+
   type 'a t =
-    { ideal : TransitionIdeal.t
-    ; simulation : ('a arith_term) Array.t
-    ; witness : TransitionIdeal.solvable_polynomial }
+    { ti : pre_t
+    ; simulation : 'a arith_term array 
+    ; constants : (symbol * symbol) list}
 
   let pp _ = assert false
 
+  let make_sp ti witness = {ideal = ti; witness = witness}
+
+  let pp_dim offset formatter i = 
+    if i = 2 * offset then 
+      Format.fprintf formatter "K"
+    else if i < offset then 
+      Format.fprintf formatter "x_%d" i
+    else
+      Format.fprintf formatter "x_%d'" (i - offset)
+
+
+  let exp_ti it = 
+    let it_offset = it.ideal.dim in
+    logf  "Exponentiating : %a" (TransitionIdeal.pp (pp_dim it_offset)) it.ideal;
+    if I.generators (it.ideal.ideal) = [] then
+      TransitionIdeal.make it_offset it.ideal.ideal
+    else if QQXs.is_zero (I.reduce it.ideal.ideal QQXs.one) then
+      TransitionIdeal.make it_offset it.ideal.ideal
+    else
+      let inv_seq, inv_dom = TransitionIdeal.iteration_sequence it.ideal in
+      let inv_dom_id = Id.make (I.generators inv_dom) in
+      let k_equal_i i = QQXs.sub (QQXs.of_dim (2 * it.ideal.dim)) (QQXs.scalar (QQ.of_int i)) in
+      let zeroth = Id.make ((k_equal_i 0) :: (List.init it_offset (fun d -> QQXs.sub (QQXs.of_dim (d+it_offset)) (QQXs.of_dim d)))) in
+      let inv_seq_id = List.mapi (fun i (id : TransitionIdeal.t) -> Id.make ((k_equal_i (i+1)) :: (I.generators id.ideal))) inv_seq in
+      let transient_closure = List.fold_left Id.intersect zeroth inv_seq_id in
+      logf "Invariant Dom : %a" (I.pp (pp_dim it_offset)) inv_dom;
+      if QQXs.is_zero (I.reduce inv_dom QQXs.one) then 
+        TransitionIdeal.make it_offset (Id.mk_rewrite transient_closure)
+      else      
+        (log_pp  pp_sp it.witness;
+        let cf = Log.time "Rat Exp" (Rational.RatEP.solve_rec) it.witness in
+        let pp_cl f = Array.iteri (fun i cl -> Format.fprintf f "cl.(%d) = %a@." i Rational.RatEP.pp cl) in
+        logf  "%a" pp_cl cf;
+        let sp_map_offset = Array.length cf in
+        let module EP = (val Log.time "Splitting Field" Rational.RatEP.to_nf cf) in
+
+        let zero_eig_transient, zero_eigen_stab, rels = Log.time "Algebraic Relations" EP.long_run_algebraic_relations () in
+        logf  "Alg Relations: %a" (Id.pp (pp_dim it_offset)) (Id.make rels);
+        
+        let cl = 
+          if (List.length inv_seq_id) + 1 >= zero_eigen_stab then
```

Comment:
Not sure that I'm reading correctly, but my understanding of this is that you want the transient length to be the max of the iteration at which the domain of ti converges and the size of the largest jordan block of 0.  That seems plausible, but I'm not sure: the solvability witness approximates the behavior of ti + dom*(ti), but not (necessarily) ti.

Consider x' = x /\ y' = x /\ y = 0 /\ z_1' = xz_1 /\ z_2' = z_1 /\ ... /\ z_n' = z_{n-1}
This transition ideal has invariant domain <x, y> (converging @ iteration 2) and solvability witness
f(x,y,z_1,...,z_n) = (0, 0, 0, z_1, ... , z_{n-1} )
(with a Jordan block of size n).  My thought is that we need a transient length of n+2 rather than max(n,2).

Reply:
I'm not sure I understand your point with the example. Let n = 4 for the above example. At the point in the code referenced we have:

 - inv_dom_id = <x, y>
 - transient_closure = ti^0 \cap ti^1 \cap ti^2
 - zero_eig_transient = [{x, y, z_1, z_2, z_3, z_4}, {0, 0, 0, z_1, z_2, z_3},  {0, 0, 0, 0, z_1, z_2},  {0, 0, 0, 0, 0, z_1}]
 - rels = [x', y', z_1', z_2', z_3', z_4']
 - zero_eigen_stab = 4

We then peel off elements of zero_eig_transient until we exceed iterations past the transient. The final answer is
  transient_closure \cap (<x', y', z_1', z_2', z_3', z_4' - z_1> + inv_dom_id) \cap (<x', y', z_1', z_2', z_3', z_4'> + inv_dom_id)

The middle term comes from the last element of zero_eig_transient. We only ever use the witness to approximate ti past the point of invariant domain stabilizations. For all iterations prior we take elements from the invariant domain sequence, which always is sound with respect to ti.

I've added this as a test case in test_transitionIdeal.ml. It's ultsolvable_cl2. The result is correct at least for the first 10 iterations of ti.